File: sections/64_multiplicative_number_theory/problems/d_squarefree_rank/solution.py
import math
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    queries = data[1 : 1 + data[0]]
    limit = math.isqrt(3 * max(queries)) + 1

    mu = [0] * (limit + 1)
    composite = bytearray(limit + 1)
    primes: list[int] = []
    mu[1] = 1
    for value in range(2, limit + 1):
        if not composite[value]:
            primes.append(value)
            mu[value] = -1
        for prime in primes:
            product = value * prime
            if product > limit:
                break
            composite[product] = 1
            if value % prime == 0:
                mu[product] = 0
                break
            mu[product] = -mu[value]

    def count_squarefree(x: int) -> int:
        return sum(
            mu[d] * (x // (d * d))
            for d in range(1, math.isqrt(x) + 1)
        )

    answers = []
    for rank in queries:
        low, high = 1, 3 * rank
        while low < high:
            middle = (low + high) // 2
            if count_squarefree(middle) >= rank:
                high = middle
            else:
                low = middle + 1
        answers.append(str(low))
    print("\n".join(answers))


if __name__ == "__main__":
    main()

File: sections/64_multiplicative_number_theory/problems/d_squarefree_rank/solution.cpp
#include <bits/stdc++.h>
using namespace std;

using int64 = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int q;
    if (!(cin >> q)) return 0;
    vector<int64> queries(q);
    int64 largest = 0;
    for (int64 &k : queries) {
        cin >> k;
        largest = max(largest, k);
    }

    int limit = sqrtl(3.0L * largest) + 2;
    vector<int> mu(limit + 1), primes;
    vector<bool> composite(limit + 1);
    mu[1] = 1;
    for (int value = 2; value <= limit; ++value) {
        if (!composite[value]) {
            primes.push_back(value);
            mu[value] = -1;
        }
        for (int prime : primes) {
            if (1LL * value * prime > limit) break;
            composite[value * prime] = true;
            if (value % prime == 0) {
                mu[value * prime] = 0;
                break;
            }
            mu[value * prime] = -mu[value];
        }
    }

    auto count_squarefree = [&](int64 x) {
        int64 result = 0;
        for (int64 d = 1; d * d <= x; ++d)
            result += mu[d] * (x / (d * d));
        return result;
    };

    for (int64 k : queries) {
        int64 low = 1, high = 3 * k;
        while (low < high) {
            int64 middle = low + (high - low) / 2;
            if (count_squarefree(middle) >= k)
                high = middle;
            else
                low = middle + 1;
        }
        cout << low << '\n';
    }
}
